[SPARK-12810][PySpark] PySpark CrossValidatorModel should support avgMetrics

vectorijk opened the pull request:
## What changes were proposed in this pull request?

support avgMetrics in CrossValidatorModel with Python
## How was this patch tested?

Doctest and `test_save_load` in `pyspark/ml/test.py`
[JIRA](https://issues.apache.org/jira/browse/SPARK-12810)

vectorijk committed:
supporting avgMetrics in CrossValidatorModel with Python

vectorijk commented:
cc @feynmanliang @jkbradley @mengxr

vectorijk commented:
Jenkins, test this please.

SparkQA commented:
**[Test build #56069 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56069/consoleFull)** for PR 12464 at commit [`93a43bc`](https://github.com/apache/spark/commit/93a43bc5007acbc9c55a3eaf591f2b16df614c68).

SparkQA commented:
**[Test build #56069 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56069/consoleFull)** for PR 12464 at commit [`93a43bc`](https://github.com/apache/spark/commit/93a43bc5007acbc9c55a3eaf591f2b16df614c68).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56069/
Test PASSed.

jkbradley commented:
I'll take a look

jkbradley commented on python/pyspark/ml/tuning.py:
```diff
@@ -367,7 +368,9 @@ def copy(self, extra=None):
             extra = dict()
-        return CrossValidatorModel(self.bestModel.copy(extra))
+        bestModel = self.bestModel.copy(extra)
+        avgMetrics = [am.copy(extra) for am in self.avgMetrics]
```
You're having to convert avgMetrics to a list since it's stored as a numpy.ndarray in CrossValidator.  Could you update CrossValidator itself so that it just uses a list of floats, rather than a numpy array?

vectorijk replied:
Sure, I will do this.

jkbradley commented on python/pyspark/ml/tuning.py:
```diff
@@ -394,9 +397,10 @@ def _from_java(cls, java_stage):
 
         # Load information from java_stage to the instance.
         bestModel = JavaParams._from_java(java_stage.bestModel())
+        avgMetrics = [am for am in java_stage.avgMetrics()]
```
Simpler: ```list(java_stage.avgMetrics())```

jkbradley commented:
That's all for now, thanks!

holdenk commented on python/pyspark/ml/tests.py:
```diff
@@ -534,6 +534,8 @@ def test_save_load(self):
         cvModel.save(cvModelPath)
         loadedModel = CrossValidatorModel.load(cvModelPath)
         self.assertEqual(loadedModel.bestModel.uid, cvModel.bestModel.uid)
+        for index in range(len(loadedModel.avgMetrics)):
```
Minor possible suggestion, there are some other places in the doctests where we use numpys assert_almost_equal, it seems like that might simplify things here a bit if you wanted to.

vectorijk replied:
@holdenk Thanks for suggestion. I used `assert_almost_equal` [here](https://github.com/vectorijk/spark/blob/spark-12810/python/pyspark/ml/tests.py#L561).

vectorijk committed:
address comment

- update metrics to list of floats
- use `numpy.testing.assert_almost_equal` to assert float list
- test CrossValidator and CrossValidatorModel copy

SparkQA commented:
**[Test build #57121 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57121/consoleFull)** for PR 12464 at commit [`25959e5`](https://github.com/apache/spark/commit/25959e546e3efe6e6b1c16c805f16e2045bc440f).

vectorijk commented:
@jkbradley 25959e5 this commit is trying to
- update `metrics` in `CrossValidator` to float list (like [0.0] * number) .
- use `numpy.testing.assert_almost_equal` to assert two float lists.
- test `CrossValidator` and `CrossValidatorModel` with copy

Do you mind review again?

SparkQA commented:
**[Test build #57121 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57121/consoleFull)** for PR 12464 at commit [`25959e5`](https://github.com/apache/spark/commit/25959e546e3efe6e6b1c16c805f16e2045bc440f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57121/
Test PASSed.

vectorijk commented:
I also notice that `validationMetrics` in `TrainValidationSplitModel` should also be supported in Python.
Should we support that after this PR?

jkbradley commented on python/pyspark/ml/tests.py:
```diff
@@ -461,6 +462,29 @@ def _fit(self, dataset):
+
+        cvModel = cv.fit(dataset)
+        cvModelCopied = cvModel.copy()
+        assert_almost_equal(cvModel.avgMetrics, cvModelCopied.avgMetrics)
```
On second thought, I'd use ```self.assertEqual``` for this and the avgMetrics comparison below.  We should know if copying or saving/loading causes loss of precision.

jkbradley commented on python/pyspark/ml/tuning.py:
```diff
@@ -346,10 +346,11 @@ class CrossValidatorModel(Model, ValidatorParams, MLReadable, MLWritable):
     """
 
-    def __init__(self, bestModel):
+    def __init__(self, bestModel, avgMetrics):
```
I just realized we should add a default value for avgMetrics since it is technically a public API.  How about a value of ```[]``` by default?

jkbradley commented:
Thanks for the updates.  I just had 2 small comments.

> I also notice that validationMetrics in TrainValidationSplitModel should also be supported in Python.
Should we support that after this PR?

Yes please, can you create a JIRA?

vectorijk committed:
address comment

SparkQA commented:
**[Test build #57248 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57248/consoleFull)** for PR 12464 at commit [`cfe6a66`](https://github.com/apache/spark/commit/cfe6a665b0c7d33b28261bac4c45fa24abe09e98).

SparkQA commented:
**[Test build #57248 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57248/consoleFull)** for PR 12464 at commit [`cfe6a66`](https://github.com/apache/spark/commit/cfe6a665b0c7d33b28261bac4c45fa24abe09e98).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57248/
Test FAILed.

SparkQA commented:
**[Test build #57249 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57249/consoleFull)** for PR 12464 at commit [`51b412f`](https://github.com/apache/spark/commit/51b412fb455b6271460cf274cc45de73a1aec885).

SparkQA commented:
**[Test build #57249 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57249/consoleFull)** for PR 12464 at commit [`51b412f`](https://github.com/apache/spark/commit/51b412fb455b6271460cf274cc45de73a1aec885).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57249/
Test PASSed.

jkbradley commented on python/pyspark/ml/tests.py:
```diff
@@ -461,6 +461,31 @@ def _fit(self, dataset):
+        cvModel = cv.fit(dataset)
+        cvModelCopied = cvModel.copy()
+        for index in range(len(cvModel.avgMetrics)):
+            self.assertTrue(abs(cvModel.avgMetrics[index] - cvModelCopied.avgMetrics[index])
```
Did you try assertEqual and find it did not work?  Why do we need approximate equality here?

vectorijk replied:
I have tried `assertEqual` before. This test case causes loss of precision under python2 if we use `assertEqual`. But under python3, it passes.

jkbradley replied:
Interesting.  OK thanks for checking!

jkbradley commented on python/pyspark/ml/tests.py:
```diff
@@ -534,6 +559,8 @@ def test_save_load(self):
         loadedModel = CrossValidatorModel.load(cvModelPath)
         self.assertEqual(loadedModel.bestModel.uid, cvModel.bestModel.uid)
+        for index in range(len(loadedModel.avgMetrics)):
+            self.assertTrue(abs(loadedModel.avgMetrics[index] - cvModel.avgMetrics[index]) < 0.0001)
```
Same here; why approximate equality?

jkbradley commented:
LGTM
Merging with master
Thanks!
